### ISSUE TITLE
rism3d.snglpnt links to MPI libraries

### ISSUE THREAD

**Issue by tluchko** [Amber-MD/cmake-buildscripts#110]
When configured with '-DMPI=True', rism3d.snglpnt is built against the MPI libraries instead of the serial libraries. This does not happen when '-DMPI=False'.

The steps to reproduce are:

`CC=gcc CXX=g++ cmake ../amber18 -DCMAKE_INSTALL_PREFIX=../amber18rc3 -DCOMPILER=GNU -DDOWNLOAD_MINICONDA=True -DINSTALL_TESTS=True -DMPI=True
cd AmberTools/src/rism
make VERBOSE=1 rism3d.snglpnt`

The link stage of the build is

`/opt/local/bin/gcc  -Wall -Wno-unused-function -Wno-unknown-pragmas -Wno-unused-variable -Wno-unused-but-set-variable -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/rism3d.snglpnt.dir/CMakeFiles/nabfiles/rism3d.snglpnt.c.o  -o rism3d.snglpnt -Wl,-rpath,/Users/tluchko/projects/testing/build-amber18/AmberTools/src/cifparse -Wl,-rpath,/Users/tluchko/projects/testing/build-amber18/AmberTools/src/pbsa -Wl,-rpath,/Users/tluchko/projects/testing/build-amber18/AmberTools/src/fftw-3.3 -Wl,-rpath,/Users/tluchko/projects/testing/build-amber18/AmberTools/src/netcdf-fortran-4.4.4/fortran -Wl,-rpath,/Users/tluchko/projects/testing/build-amber18/AmberTools/src/netcdf-4.3.0/liblib -Wl,-rpath,/Users/tluchko/projects/testing/build-amber18/AmberTools/src/arpack ../nab/libnab_mpi.a ../cifparse/libcifparse.dylib ../sff/libsff_mpi.a librism.a ../xblas/build/libxblas-amb.a -Wl,-flat_namespace /opt/local/lib/mpich-gcc5/libmpifort.dylib /opt/local/lib/mpich-gcc5/libmpi.dylib /opt/local/lib/mpich-gcc5/libpmpi.dylib ../pbsa/libpbsa.dylib ../fftw-3.3/libfftw3.dylib ../lib/libamber_common.a ../netcdf-fortran-4.4.4/fortran/libnetcdff.6.1.1.dylib ../netcdf-4.3.0/liblib/libnetcdf.dylib /usr/lib/libz.dylib ../arpack/libarpack.dylib -framework Accelerate -lgfortran -lquadmath -lm `

When I build with '-DMPI=False', rism3d.snglpnt links correctly.

I've attached debug files for the -DMPI=True case.

[errorreport.tar.gz](https://github.com/Amber-MD/cmake-buildscripts/files/1862163/errorreport.tar.gz)

**Comment by multiplemonomials** [comment]
Found it, it was a copy paste error in `nab.cmake`.  I'll push a commit to fix it soon.

**Comment by multiplemonomials** [comment]
Fixed by cc0d8dd5da8e9e

**Comment by tluchko** [comment]
It seems this didn't quite work.  The serial build works fine now but a whole bunch of 3D-RISM tests fail with the MPI build. When run, it seems that each process of rism3d.snglpnt.MPI, for example, thinks it is the only process.  It is a little odd because one process gets to read input files while the others don't but they all have rank 0.  

I can confirm that this doesn't happen when I use the old configure script and all the tests pass.

The executable is linking to the MPI libraries, so I'm not sure what the problem is.

[makeoutput.log](https://github.com/Amber-MD/cmake-buildscripts/files/1909887/makeoutput.log)

**Comment by swails** [comment]
Sounds like it’s linking the wrong MPI libs. I used to see this all the time when mpirun came from a different MPI installation than the compiler wrappers used to build amber. 

Can you use ldd to determine which MPI libraries the binaries are linked to and make sure it’s the same MPI from which you’re using mpirun?

**Comment by tluchko** [comment]
This is on a Mac (I can't get past cmake on my cluster) so there is no ldd but there otool:

>otool -L rism3d.snglpnt.MPI
rism3d.snglpnt.MPI:
	@rpath/libcifparse.dylib (compatibility version 0.0.0, current version 0.0.0)
	/opt/local/lib/mpich-gcc5/libmpifort.12.dylib (compatibility version 14.0.0, current version 14.1.0)
	/opt/local/lib/mpich-gcc5/libmpi.12.dylib (compatibility version 14.0.0, current version 14.1.0)
	/opt/local/lib/mpich-gcc5/libpmpi.12.dylib (compatibility version 14.0.0, current version 14.1.0)
	@rpath/libpbsa.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libfftw3.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libnetcdff.6.dylib (compatibility version 6.0.0, current version 6.1.1)
	@rpath/libnetcdf.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	@rpath/libarpack.dylib (compatibility version 0.0.0, current version 0.0.0)
	/System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
	/opt/local/lib/libgcc/libgfortran.3.dylib (compatibility version 4.0.0, current version 4.0.0)
	/opt/local/lib/libgcc/libquadmath.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
	/opt/local/lib/libgcc/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)


I only had mpich-gcc5-fortran installed at the time and it seems that is what it is picking up.

**Comment by multiplemonomials** [comment]
@tluchko Do you mind taking a full bug report on this system?  I need to take a look at the CMake cache files.

**Comment by tluchko** [comment]
Sorry for the long wait.  This fell off my radar.  Please see attached.

[errorreport.tar.gz](https://github.com/Amber-MD/cmake-buildscripts/files/1968386/errorreport.tar.gz)

**Comment by multiplemonomials** [comment]
Can you post the output of `which mpirun`?  Also, if you have a rism3d.snglpnt.MPI executable that was built using the old configure script, and post the output of running otool on that?  Thanks for your help!

**Comment by multiplemonomials** [comment]
One more think, I just discovered that in your earlier bug report, MPI was turned on for the `cmakeoutput.log`, but the `makeoutput.log` is for a serial build, in fact I can't even find the string MPI in the entire file.  Any idea why this would have happened?

**Comment by tluchko** [comment]
> Can you post the output of which mpirun? 

```
>which mpirun
/opt/local/bin/mpirun
```
> Also, if you have a rism3d.snglpnt.MPI executable that was built using the old configure script, and post the output of running otool on that?

```
>otool -L rism3d.snglpnt.MPI
rism3d.snglpnt.MPI:
	/Users/tluchko/projects/testing/master-ln/lib/libfftw3_mpi.3.dylib (compatibility version 7.0.0, current version 7.0.0)
	/Users/tluchko/projects/testing/master-ln/lib/libfftw3.3.dylib (compatibility version 7.0.0, current version 7.0.0)
	/Users/tluchko/projects/testing/master-ln/lib/libnetcdf.7.dylib (compatibility version 10.0.0, current version 10.0.0)
	/opt/local/lib/libgcc/libgfortran.3.dylib (compatibility version 4.0.0, current version 4.0.0)
	/System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
	/opt/local/lib/mpich-gcc5/libmpifort.12.dylib (compatibility version 14.0.0, current version 14.1.0)
	/opt/local/lib/mpich-gcc5/libmpi.12.dylib (compatibility version 14.0.0, current version 14.1.0)
	/opt/local/lib/mpich-gcc5/libpmpi.12.dylib (compatibility version 14.0.0, current version 14.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.50.4)
	/opt/local/lib/libgcc/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
```

> One more think, I just discovered that in your earlier bug report, MPI was turned on for the cmakeoutput.log, but the makeoutput.log is for a serial build, in fact I can't even find the string MPI in the entire file. Any idea why this would have happened?

'Cause I messed up. I sent you output for building rism3d.snglpnt instead of rism3d.snglpnt.MPI.  The correct makoutput.log is attached.

[makeoutput.log](https://github.com/Amber-MD/cmake-buildscripts/files/1975725/makeoutput.log)

**Comment by multiplemonomials** [comment]
OK I think I found the issue.  It turned out that `rism3d.snglpnt.MPI` was accidentally getting linked to serial librism through libsff!  I fixed it so that libsff_mpi now links to librism_mpi, and things seem to be working now.   

Can you check if this is fixed by Amber master commit 28860f947ecb26e?

**Comment by tluchko** [comment]
Looks like this did the trick.  Thanks!
